### PR TITLE
add slurm tests for cli_filter and task affinity

### PR DESCRIPTION
**Pull request by smoors**

### CODE REVIEW

File: slurm/slurm.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+
+os.environ["TEST_ENVAR_OUTSIDE"] = 'defined'
+
+affinity_script = """
+import os
+affinity = os.sched_getaffinity(0)
+print("num_cores: " + str(len(affinity)))
+for aff in affinity:
+    print("affinity: " + str(aff))"""
```
Comment (wpoely86) [inline slurm/slurm.py]: can't we better let this spew out json?

File: slurm/slurm.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    num_tasks = 1
+    num_tasks_per_node = 1
+    num_cpus_per_task = 1
+
+
+@rfm.simple_test
+class SbatchCleanEnvTest(rfm.RunOnlyRegressionTest):
+    descr += "sbatch starts in a clean environment"
+    exe = 'print(os.getenv("TEST_ENVAR_OUTSIDE") is None)'
+    executable = f"python3 -c 'import os;{exe}'"
```
Comment (wpoely86) [inline slurm/slurm.py:43]: why split this in two strings?

File: slurm/slurm.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    @sanity_function
+    def assert_env(self):
+        return sn.assert_found(r'^True$', self.stdout, self.descr)
```
Comment (wpoely86) [inline slurm/slurm.py]: should we also check that stderr is empty?

File: slurm/slurm.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+
+@rfm.simple_test
+class SbatchSrunCopyEnvTest(rfm.RunOnlyRegressionTest):
+    descr += "srun copies the sbatch job environment"
+    prerun_cmds = ['export TEST_ENVAR_INSIDE=defined']
```
Comment (wpoely86) [inline slurm/slurm.py]: can we insert something random to be sure? Like a datestring?

File: slurm/slurm.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    exe = 'print(os.environ["TEST_ENVAR_INSIDE"] == "defined")'
```
Comment (wpoely86) [inline slurm/slurm.py]: use a get here

File: slurm/slurm.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    executable = f"srun python3 -c 'import os;{exe}'"
+
+    @sanity_function
+    def assert_env(self):
+        return sn.assert_found(r'^True$', self.stdout, self.descr)
+
+
+@rfm.simple_test
+class SbatchEnforceBindingTest(rfm.RunOnlyRegressionTest):
```
Comment (wpoely86) [inline slurm/slurm.py]: can we also check that is actually works (in the gpu case)?

Comment (smoors) [reply slurm/slurm.py]: that's planned for another test

File: slurm/slurm.py (hunk 1 above)
Comment (smoors) [reply slurm/slurm.py]: using json is a bit more work to implement.
the thing is that you can have 1 or more tasks running this code, so each task has to write its own json file, and the check has to read each of them separately.

on the other hand, each task writing its own file is cleaner than lumping all output into the log file.

File: slurm/slurm.py (hunk 2 above)
Comment (smoors) [reply slurm/slurm.py:43]: it looks cleaner to me this way, but i'll merge it if you prefer

File: slurm/slurm.py [hunk 7]
```diff
@@ -1,16 +1,29 @@
...
+affinity = list(os.sched_getaffinity(0))
+task_id = os.getenv("SLURM_PROCID", "0") + "_" + os.getenv("SLURM_STEP_ID", "0")
+json.dump(affinity, open(f"affinity{task_id}.json", "w"))
+time.sleep(20)
+"""
+
+
+def load_affinities():
+    affinities = []
+    for aff in glob.glob('affinity*.json'):
```
Comment (wpoely86) [inline slurm/slurm.py:24]: and this happens in per test/per job directory so we're sure it's only the just created `affinity.....json` files?

Comment (smoors) [reply slurm/slurm.py:24]: yes, each test runs in its own subdirectory under `stage` and this dir is automatically cleaned up before re-running the test.

**Review by wpoely86 — APPROVED** [review]

**Pull request merged**